### PR TITLE
net/tcp_obstructed_close

### PR DESCRIPTION
**Pull request by ssharks**
A test to reproduce the issue as mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/44507.

When a connection is closed, but the other side does not ACK the FIN message, no retry is done

### CODE REVIEW

File: tests/net/socket/tcp/src/main.c [hunk 1]
```diff
@@ -723,6 +723,108 @@ void test_open_close_immediately(void)
...
+	int dropped_packets = dropped_packets_after - dropped_packets_before;
+
+	/* At least some packet should have been dropped */
+	zassert_equal(dropped_packets,
+			CONFIG_NET_TCP_RETRY_COUNT,
+			"Insuffcient FIN retries, got %i", dropped_packets);
+
+	start_time = k_uptime_get_32();
+
+	/* After the client socket closing, the context count should be 1 */
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:780]: Actually, this is also not true, we still have two sockets allocated at this point (the listening `s_socks_sock and `new_sock` pending disconnection).

File: tests/net/socket/tcp/src/main.c [hunk 2]
```diff
@@ -723,6 +723,108 @@ void test_open_close_immediately(void)
...
+	zassert_equal(connect(c_sock, (struct sockaddr *)&s_saddr,
+				sizeof(s_saddr)),
+				0, "connect not succeed");
+	test_accept(s_sock, &new_sock, &addr, &addrlen);
+
+	/* We should have two contexts open now */
+	net_context_foreach(calc_net_context, &count_before);
+
+	/* Break the communication */
+	loopback_set_packet_drop_ratio(1.0f);
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:762]: If we want to verify if FIN is actually retransmitted, I suggest a simpler test:
```
	/* Break the communication */
	loopback_set_packet_drop_ratio(1.0f);

	test_close(c_sock);
   
	k_sleep(K_MSEC(200));

	/* Server socket should not notify connection closed at this point yet */
	ret = recv(new_sock, &buf, sizeof(buf), MSG_DONTWAIT);
	zassert_equal(ret, -1, "not -1");
	zassert_equal(errno, EAGAIN, "not EAGAIN");

	loopback_set_packet_drop_ratio(0.0f);
	/* Give the retransmission timer a chance to kick in */
	k_sleep(K_MSEC(500));

	/* Server socket should report closed connection now */
	ret = recv(new_sock, &buf, sizeof(buf), MSG_DONTWAIT);
	zassert_equal(ret, 0, "not 0");

	test_close(new_sock);
	test_close(s_sock);
```

Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:762]: Thanks for the suggestion. Looking at this, we might want to make 2 different tests. The existing test to validate the proper closure of the context, the other to count check the FIN retry as you suggested. Do you agree?

Comment (rlubos) [reply tests/net/socket/tcp/src/main.c:762]: net context close is already verified by `test_open_close_immediately`, I don't think if there's a need to duplicate that.

**Comment by ssharks** [comment]
After moving the check on the number of retransmissions in the test I get 5 retransmissions instead of the 9 set in the CONFIG_NET_TCP_RETRY_COUNT. This likely has to do with the retransmission timeout being set to 200 ms, and the socket close timer to 200 ms.
Do we accept this behavior, by modifying the test?

**Comment by rlubos** [comment]
> After moving the check on the number of retransmissions in the test I get 5 retransmissions instead of the 9 set in the CONFIG_NET_TCP_RETRY_COUNT. This likely has to do with the retransmission timeout being set to 200 ms, and the socket close timer to 200 ms. Do we accept this behavior, by modifying the test?

I'm thinking, maybe instead of using a fixed `FIN_TIMEOUT` set to 1 second, we could calculate it's value based on the other, configurable parameters? The maximum time needed to send all the retransmissions i a function of `CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT` and `CONFIG_NET_TCP_RETRY_COUNT`, we could set the `FIN TIMEOUT` to be `CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT * (CONFIG_NET_TCP_RETRY_COUNT + 1)`?

**Comment by ssharks** [comment]
> I'm thinking, maybe instead of using a fixed `FIN_TIMEOUT` set to 1 second, we could calculate it's value based on the other, configurable parameters? The maximum time needed to send all the retransmissions i a function of `CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT` and `CONFIG_NET_TCP_RETRY_COUNT`, we could set the `FIN TIMEOUT` to be `CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT * (CONFIG_NET_TCP_RETRY_COUNT + 1)`?

That would make sense, the current solution is very prone to errors, because there are actually 3 values determining 2 things. The situation is over determined. Be aware that the rto can be changed runtime, so this means the FIN_TIMEOUT should be computed runtime as well.

File: tests/net/socket/tcp/src/main.c [hunk 3]
```diff
@@ -723,6 +723,108 @@ void test_open_close_immediately(void)
...
+
+	dropped_packets_before = loopback_get_num_dropped_packets();
+
+	test_close(c_sock);
```
[the remark this answers is not in the document]
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:766]: Fixed by moving the test to below where the context has been removed

File: tests/net/socket/tcp/src/main.c (hunk 1 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:770]: You are right, change to: After the client socket closing, the context count should be 1 less

File: tests/net/socket/tcp/src/main.c (hunk 2 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c:762]: I think it is better to have both the context closing and counting of FIN packets checked

File: tests/net/socket/tcp/src/main.c [hunk 4]
```diff
@@ -723,6 +723,106 @@ void test_open_close_immediately(void)
...
+	zassert_equal(count_before - 1, count_after,
+		      "net_context still in use (before %d vs after %d)",
+		      count_before - 1, count_after);
+
+	dropped_packets_after = loopback_get_num_dropped_packets();
+	int dropped_packets = dropped_packets_after - dropped_packets_before;
+
+	/* At least some packet should have been dropped */
+	zassert_equal(dropped_packets,
+			CONFIG_NET_TCP_RETRY_COUNT,
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:792]: Just noticed that this should actually be `CONFIG_NET_TCP_RETRY_COUNT + 1` - plus 1 for initial transmission.

**Comment by rlubos** [comment]
> Be aware that the rto can be changed runtime, so this means the FIN_TIMEOUT should be computed runtime as well.

Atually, I couldn't find where would this be changed runtime, but since it's actually stored in a variable we should probably consider that. The FIN timeout could be computed based on `tcp_rto` and `tcp_retries` values. Will you add a respective change in the TCP stack in order to make the test pass in this PR (shoud be a simple change), or do you want me to do this in another PR?

**Comment by ssharks** [comment]
A fix for the number of FIN messages is proposed by https://github.com/zephyrproject-rtos/zephyr/pull/45103

**Comment by ssharks** [comment]
@rlubos: Could you have a look at the current status?

File: tests/net/socket/tcp/src/main.c [hunk 5]
```diff
@@ -723,6 +723,107 @@ void test_open_close_immediately(void)
...
+	zassert_equal(count_before - 1, count_after,
+		      "net_context still in use (before %d vs after %d)",
+		      count_before - 1, count_after);
+
+	dropped_packets_after = loopback_get_num_dropped_packets();
+	int dropped_packets = dropped_packets_after - dropped_packets_before;
+
+	/* At least some packet should have been dropped */
+	zassert_equal(dropped_packets,
+			CONFIG_NET_TCP_RETRY_COUNT+1,
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:792]: nit
```suggestion
			CONFIG_NET_TCP_RETRY_COUNT + 1,
```

**Review by rlubos — DISMISSED** [review]
Thanks, LGTM

File: tests/net/socket/tcp/src/main.c (hunk 5 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: Fixed

File: tests/net/socket/tcp/src/main.c (hunk 4 above)
Comment (ssharks) [reply tests/net/socket/tcp/src/main.c]: Fixed

**Review by rlubos — DISMISSED** [review]

**Comment by ssharks** [comment]
@jukkar Could you have a look at this one?

**Review by jukkar — DISMISSED** [review]

**Comment by rlubos** [comment]
@ssharks There is a conflict in the test file.

**Comment by ssharks** [comment]
> @ssharks There is a conflict in the test file.

Thanks, did not notice that, I did a manual merge to fix the issue

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**